### PR TITLE
Fix probability computation in `WhisperNoSpeechDetection` when recomputing scores

### LINKED ISSUE #29313: Incorrect no-speech probability in `WhisperNoSpeechDetection` when `language` is set and `num_beams > 1`
### System Info

- `transformers` version: 4.39.0.dev0 (`c8d9840`)

### Who can help?

@patrickvonplaten @sanchit-gandhi

### Reproduction

 In [`WhisperNoSpeechDetection`](https://github.com/huggingface/transformers/blob/c8d98405a8f7b0e5d07391b671dcc61bb9d7bad5/src/transformers/generation/logits_process.py#L1893), in some cases, the model is run again inside the logits processor to get the **logits**, but then, if `self.is_scores_logprobs == True`, those logits are run through `exp()` instead of `softmax()`, resulting in incorrect (unnormalized) probabilities (possibly `> 1`).

This scenario arises when the model is called with `language` set and with `num_beams > 1`.

### Expected behavior

The no-speech probability should be the same regardless of the value of `num_beams`.

### PR DESCRIPTION
**Pull request by cifkao**
# What does this PR do?

Fix #29313.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@patrickvonplaten @sanchit-gandhi @ylacombe

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
For whisper, `num_beams > 1` paths seems a bit tricky. Would you mind adding a test to make sure we have the expected new results?

**Comment by cifkao** [comment]
@ArthurZucker Added a slow test where I set the `logprob_threshold` to 0 to make sure the no-speech detection is triggered. Without this fix, all the outputs are empty because all the no-speech probabilities are >1.

**Comment by ylacombe** [comment]
Hey @cifkao, thanks for the PR!
Could you make sure to rebase the branch to fix the failing test ?

> This scenario arises when the model is called with language set and with num_beams > 1.

Also, could you point out why this scenario applies when language is set ? I understand the case for num_beams>1 but don't see the point for the other case!
Many thanks!

**Comment by cifkao** [comment]
> Also, could you point out why this scenario applies when language is set ? I understand the case for num_beams>1 but don't see the point for the other case!

The bug manifests only when _both_ conditions (`num_beams > 1` and `language` set) are true. The latter condition causes decoding to start after the language token, which is why we need to call the model again to get the logits at that position (which are then incorrectly treated as log-probabilities). With `language` unset, the decoding starts from the beginning, so we never enter that branch.

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -1954,17 +1954,20 @@ def set_begin_index(self, begin_index):
...
+        is_scores_logprobs = self.is_scores_logprobs
+
         if input_ids.shape[1] == self.begin_index:
             if self.start_of_trans_offset > 1:
                 with torch.no_grad():
                     logits = self.model(**self.inputs).logits
 
                 no_speech_index = self.begin_index - self.start_of_trans_offset
                 no_speech_scores = logits[:, no_speech_index]
+                is_scores_logprobs = False
```
Comment (sanchit-gandhi) [inline src/transformers/generation/logits_process.py:1966]: Nice catch!

File: tests/models/whisper/test_modeling_whisper.py [hunk 2]
```diff
@@ -2615,6 +2615,59 @@ def test_whisper_longform_multi_batch_hard_prev_cond(self):
         for i in range(num_samples):
             assert decoded_all[i] == EXPECTED_TEXT[i]
 
+    @slow
+    def test_whisper_longform_no_speech_detection(self):
```
Comment (sanchit-gandhi) [inline tests/models/whisper/test_modeling_whisper.py:2619]: Thanks for adding this slow test! Just to confirm, before the fix all the transcriptions are empty due to the no-speech probabilities exceeding 1?

**Review by sanchit-gandhi — APPROVED** [review]
Thanks for this fix @cifkao!

**Comment by sanchit-gandhi** [comment]
Ready for final review from @ArthurZucker

File: tests/models/whisper/test_modeling_whisper.py (hunk 2 above)
Comment (cifkao) [reply tests/models/whisper/test_modeling_whisper.py:2619]: Exactly.

**Review by ArthurZucker — APPROVED** [review]
🤯 awesome catch!

**Comment by ArthurZucker** [comment]
Thank you for taking the time to add a test 🔥

**Comment by sanchit-gandhi** [comment]
Thanks for the contribution @cifkao!
